homeassistant/components/light/template.py

```diff
@@ -0,0 +1,192 @@
+"""
+Support for Template lights.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/light.template/
+"""
+import logging
+import asyncio
+
+import voluptuous as vol
+
+from homeassistant.core import callback
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS, ENTITY_ID_FORMAT, Light, SUPPORT_BRIGHTNESS)
+from homeassistant.const import (
+    CONF_VALUE_TEMPLATE, ATTR_ENTITY_ID, ATTR_FRIENDLY_NAME, STATE_ON,
+    STATE_OFF, EVENT_HOMEASSISTANT_START
+)
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA
+from homeassistant.exceptions import TemplateError
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import async_generate_entity_id
+from homeassistant.helpers.event import async_track_state_change
+from homeassistant.helpers.restore_state import async_get_last_state
+from homeassistant.helpers.script import Script
+
+_LOGGER = logging.getLogger(__name__)
+_VALID_STATES = [STATE_ON, STATE_OFF, 'true', 'false']
+
+CONF_LIGHTS = 'lights'
+ON_ACTION = 'turn_on'
+OFF_ACTION = 'turn_off'
+LEVEL_ACTION = 'set_level'
+LEVEL_TEMPLATE = 'level_template'
+
+
+LIGHT_SCHEMA = vol.Schema({
+    vol.Required(ON_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Required(OFF_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_VALUE_TEMPLATE, default=None): cv.template,
+    vol.Optional(LEVEL_ACTION, default=None): cv.SCRIPT_SCHEMA,
+    vol.Optional(LEVEL_TEMPLATE, default=None): cv.template,
+    vol.Optional(ATTR_FRIENDLY_NAME, default=None): cv.string,
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_LIGHTS): vol.Schema({cv.slug: LIGHT_SCHEMA}),
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up Template Lights."""
+    lights = []
+
+    for device, device_config in config[CONF_LIGHTS].items():
+        friendly_name = device_config.get(ATTR_FRIENDLY_NAME, device)
+        state_template = device_config[CONF_VALUE_TEMPLATE]
+        on_action = device_config[ON_ACTION]
+        off_action = device_config[OFF_ACTION]
+        level_action = device_config[LEVEL_ACTION]
+        level_template = device_config[LEVEL_TEMPLATE]
+        entity_ids = []
```

Comment:
If no templates are defined, this list still needs to include the configured entity IDs.